### ISSUE TITLE
Angular-Electron Loading Error

### ISSUE THREAD

**Issue by bvanfleet** [TimelordUK/node-sqlserver-v8#127]
## Issue Details

I'm working on an Angular-Electron-TypeORM application. I've forked this repo and updated the electron dependency version to v7. However, I've had an issue even using the official msnodesqlv8. When Angular builds it drops the bin directory. I've gotten around this by adding an asset configuration in the `angular.json` file:

```json
"assets": [
  {
    "glob": "**/*",
    "input": "./node_modules/msnodesqlv8/lib/bin",
    "output": "./bin"
  }
]
```

## Error Details

While the application loads in debug, it's still throwing the error: `Cannot find module` (stack below):

```
Error: Cannot find module 'C:\path\to\built\app\dist\build\Debug\sqlserverv8.node'
    at webpackEmptyContext (/C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:171637)
    at Object.debugLoad (/C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:171692)
    at Object../node_modules/msnodesqlv8/lib/bootstrap.js (/C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:171734)
    at __webpack_require__ (/C:/path/to/built/app/dist/runtime-es2015.d6c52737d4587c65265f.js:80)
    at /C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:171757
    at Object../node_modules/msnodesqlv8/lib/connection.js (/C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:172193)
    at __webpack_require__ (/C:/path/to/built/app/dist/runtime-es2015.d6c52737d4587c65265f.js:80)
    at Object.<anonymous> (/C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:174237)
    at Object../node_modules/msnodesqlv8/lib/sql.js (/C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:174292)
    at __webpack_require__ (/C:/path/to/built/app/dist/runtime-es2015.d6c52737d4587c65265f.js:80)
```

If I try to use liveLoad, the error thrown is `None of the binaries loaded successfully. Is your node version >= 10.15 ?` (again, stack below)

```
Uncaught Error: None of the binaries loaded successfully. Is your node version >= 10.15 ?
    at failIfNoBinaryExported (main-es2015.1b472c6192034249cbbe.js:171719)
    at Object.liveLoad (main-es2015.1b472c6192034249cbbe.js:171706)
    at Object../node_modules/msnodesqlv8/lib/bootstrap.js (main-es2015.1b472c6192034249cbbe.js:171735)
    at __webpack_require__ (runtime-es2015.d6c52737d4587c65265f.js:80)
    at main-es2015.1b472c6192034249cbbe.js:171757
    at Object../node_modules/msnodesqlv8/lib/connection.js (main-es2015.1b472c6192034249cbbe.js:172193)
    at __webpack_require__ (runtime-es2015.d6c52737d4587c65265f.js:80)
    at Object.<anonymous> (main-es2015.1b472c6192034249cbbe.js:174237)
    at Object../node_modules/msnodesqlv8/lib/sql.js (main-es2015.1b472c6192034249cbbe.js:174292)
    at __webpack_require__ (runtime-es2015.d6c52737d4587c65265f.js:80)
```

I checked that the angular build directory contains the lib/bin directory and its contents. I also checked that it has a Build/Debug directory with the build contents. I did also verify security in the directory (just in case).

I feel as though I'm missing something simple. However, are there any considerations for using this library within an angular application? I think this may be worth a brief mention in the documentation if angular built apps have any special needs.

## Version Details

**Angular**: 8.2.12
**Electron**: 7.0.0
**Node**: 12.8.1
**msnodesqlv8**: 0.8.5 and forked with updated electron dependency

**Comment by TimelordUK** [comment]
hello, can you try latest 0.8.6 I have added electron vs 7 binaries, I tried a test seems to work ok

**Comment by bvanfleet** [comment]
Thanks for providing that release. I installed it and verified that the built angular application contained the bin directory with the new node binaries. However, I'm still getting the `None of the binaries loaded successfully. Is your node version >= 10.15 ?` error.

I'm going to be checking my Electron app build to see if it's dropping the bin directory there (I'd be surprised if it was).

I did check in the devtools debugger that `files` in liveLoad does contain all the bin directory contents, including the new Electron 7 build. I also updated my NodeJS to 12.13.0 just in case.

**Comment by bvanfleet** [comment]
I attempted to add some additional debugging lines to the bootstrap.js file, and got the following (repeated for all .node drivers):

```
Attempting to load module: 'C:\path\to\app\dist\bin\sqlserverv8.node.v12.13.0.electron.v7.1.2.x64.node'...

Unable to load module: Cannot find module 'C:\path\to\app\dist\bin\sqlserverv8.node.v12.13.0.electron.v7.1.2.x64.node'
attemptToExportBinary @ main-es2015.709db0edd2e31fa9deb4.js:171716

Error: Cannot find module 'C:\path\to\app\dist\bin\sqlserverv8.node.v12.13.0.electron.v7.1.2.x64.node'
    at webpackEmptyContext (main-es2015.709db0edd2e31fa9deb4.js:171637)
    at attemptToExportBinary (main-es2015.709db0edd2e31fa9deb4.js:171713)
    at main-es2015.709db0edd2e31fa9deb4.js:171702
    at Array.forEach (<anonymous>)
    at Object.liveLoad (main-es2015.709db0edd2e31fa9deb4.js:171700)
    at Object../node_modules/msnodesqlv8/lib/bootstrap.js (main-es2015.709db0edd2e31fa9deb4.js:171739)
    at __webpack_require__ (runtime-es2015.d6c52737d4587c65265f.js:80)
    at main-es2015.709db0edd2e31fa9deb4.js:171761
    at Object../node_modules/msnodesqlv8/lib/connection.js (main-es2015.709db0edd2e31fa9deb4.js:172197)
    at __webpack_require__ (runtime-es2015.d6c52737d4587c65265f.js:80)
```

Again, I verified that the path is valid and that the node module referenced exists at the given location. However, for some reason, it's not able to import the app. I'm curious, you mentioned getting a test working on your side, did your tsconfig file align with what I have mine set to below? I'm curious if a build config is causing a problem.

## TS Config Details
**tsconfig.json**:
```json
  "compilerOptions": {
    "outDir": "./dist/out-tsc",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "target": "es6",
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2017",
      "es2016",
      "es2015",
      "dom"
    ]
  }
```
**tsconfig.app.json**:
```json
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "outDir": "../out-tsc/app",
    "module": "esnext",
    "baseUrl": "",
    "types": []
  }
```

**Comment by TimelordUK** [comment]
If we go back to basics and install the demo hello world electron app with its main js file and you add a using msnodesqlv8 with electron v7 does that app run and start without raising the error. If it does it shows we can at least run an electron app v7. If this works we then need to start looking at what stage the failure happens.

**Comment by bvanfleet** [comment]
I agree with getting back to basics (my mind was spinning when I was troubleshooting last week) I'll test this a little later today and will report back.

**Comment by bvanfleet** [comment]
Alright, we got proof that a [super basic test](https://github.com/bvanfleet/FamilyHistoryVault) using Electron v7 works :yay: I was kind of expecting that. This means that this issue isn't necessarily related to electron, but something deeper. If I had to guess, I'd point my finger at the Angular side of it, since the angular compiler is building before it's run in electron.

**Comment by bvanfleet** [comment]
Figured it out and got it working (at least to the next error) by manually manipulating the bootstrap.js file. Webpack changes how `require` works, and it has certain limitations around how dynamic imports are called. The issue is with lines 56-57 in bootstrap.js:

```javascript
const native = path.join(binaryDir, filename)
module.exports = require(native)
```

webpack creates an empty context for that and throws an error immediately. However, if I changed line 57 to read `module.exports = require('./bin/${native}')` (which was valid due to my angular.json settings) then it works.

One item of note for Angular developers who are trying to use this with Electron, they will be required to use the angular custom webpack builder and node-loader in order for this to work successfully.

Here's an issue that mirrored the error that I was getting in webpack's repo: https://github.com/webpack/webpack/issues/8515
